macos/FluentUI/Button/Button.swift

```diff
@@ -384,129 +384,101 @@ class ButtonCell: NSButtonCell {
+		} else if yOffsetSign != 0 {
+			y += CGFloat(yOffsetSign) * (titleSize.height + titleToImageSpacing - titleToImageVerticalSpacingAdjustment) / 2
+		}
+
+		return NSMakeRect(x, y, imageSize.width, imageSize.height)
 	}
 
 	override func titleRect(forBounds rect: NSRect) -> NSRect {
-		guard let font = font else {
+		guard
+			let font = font,
+			let controlView = controlView,
+			imagePosition != .imageOnly
+		else {
 			return NSZeroRect
 		}
 
-		let isRTL = NSApp.userInterfaceLayoutDirection == .rightToLeft
-		let titleRect = super.titleRect(forBounds: rect);
+		let layoutDirectionSign = controlView.userInterfaceLayoutDirection == .rightToLeft ? -1 : 1
 		let titleSize = title.size(withAttributes: [.font: font])
 		let imageSize = image?.size ?? NSZeroSize
 
-		var x = titleRect.origin.x
-		var y = titleRect.origin.y
-		var width = titleRect.size.width
-		var height = titleRect.size.height
+		// Title is either centered, or offset from center by the image
+		var xOffsetSign = 0
+		var yOffsetSign = 0
 
 		switch imagePosition {
-		case .imageOnly:
-			x = 0
-			y = 0
-			width = 0
-			height = 0
-		case .noImage, .imageOverlaps:
-			x = horizontalPadding
-			y = verticalPadding + titleVerticalPositionAdjustment
-			width = rect.width - (horizontalPadding * 2)
-			height = rect.height - (verticalPadding * 2)
 		case .imageLeft:
-			x = horizontalPadding + imageSize.width + titleToImageSpacing
-			y = verticalPadding + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = rect.height - (verticalPadding * 2)
+			xOffsetSign = 1
 		case .imageLeading:
-			x = isRTL ? horizontalPadding : horizontalPadding + imageSize.width + titleToImageSpacing
-			y = verticalPadding + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = rect.height - (verticalPadding * 2)
+			xOffsetSign = layoutDirectionSign
 		case .imageRight:
-			x = horizontalPadding
-			y = verticalPadding + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = rect.height - (verticalPadding * 2)
+			xOffsetSign = -1
 		case .imageTrailing:
-			x = isRTL ? horizontalPadding + imageSize.width + titleToImageSpacing : horizontalPadding
-			y = verticalPadding + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = rect.height - (verticalPadding * 2)
+			xOffsetSign = -layoutDirectionSign
 		case .imageBelow:
-			x = (rect.width - titleSize.width) / 2
-			y = (rect.height - titleSize.height - imageSize.height - titleToImageSpacing) / 2 + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = titleSize.height
+			yOffsetSign = -1
 		case .imageAbove:
-			x = (rect.width - titleSize.width) / 2
-			y = (rect.height - titleSize.height + imageSize.height + titleToImageSpacing) / 2 + titleVerticalPositionAdjustment
-			width = titleSize.width
-			height = titleSize.height
+			yOffsetSign = 1
+		case .noImage, .imageOnly, .imageOverlaps:
+			// .imageOnly case is covered by guard return at top
```

Comment:
Can be preconditionFailure to enforce?

Reply:
Though does it really apply to the .noImage or imageOverlaps case?

Reply:
Similar question for imageRect

Reply:
Put cases back where they were previously and added preconditionFailure.